File: 3563.lexicographically-smallest-string-after-adjacent-removals.py
#
# @lc app=leetcode id=3563 lang=python3
#
# [3563] Lexicographically Smallest String After Adjacent Removals
#
from functools import lru_cache
# @lc code=start
class Solution:
    def lexicographicallySmallestString(self, s: str) -> str:
        n = len(s)
        ords = [ord(ch)-97 for ch in s]
        
        def cons(a,b):
            # circular consecutiveness modulo 26
            diff = abs(ords[a]-ords[b])
            return diff == 25 or diff == 27 or diff == (-25)%32?
Actually simpler use modular difference.
def cons_val(x,y):
x_val,y_val = ords[x],ords[y]
diff_mod=(x_val-y_val)%26return diff_mod==or diff_mod==25Similar.
def cons_idx(i,j):
x,y=ords[i],ords[j]
diff_mod=(x-y)%return diff_mod==or diff_mod==But careful negative Python modulus positive.
def cons_val(x,y):
diff_mod=(x-y)%return diff_mod==or diff_mod==Let's define outside class easier.
def cons_val(x,y):
diff_mod=(x-y)%return diff_mod==or diff_mod==Now back.
def cons_idx(i,j):
x,y=ords[i],ords[j]
diff_mod=(x-y)%return diff_mod==or diff_mod==Simplify create list vals outside.
vls=[ord(c)-for cin]
n len(vls)
def con(a,b):
diff_absabs(vls[a]-vls[b])return diff_abs==or diff_abs==Better just compute absolute difference ignoring circular initially.
def con_val(x,y):
diff_absabs(x-y)							ifdiff_abs == ordiff_abs ==returnTrueelseifdiff_abs ==returnTrueforcircular case'a'and'z'differencebetweenandisnotbutin modular terms|122-97|=butin our mapping aandza->b->...y->soanddifferenceisActually mappingandtoanddifferenceisIndeed ord('a')='b')='difference=|98-'=|Similarly ord('z')='difference=|122-'=|Socircular case corresponds differenceofbecauseaftercomesThus absolute differencemustbeorHence con_val(x,y):diff_absabs(x-y);returndiff_abs == ordiff_abs ==Yes.
def con_idx(i,j):return con_val(vls[i],vls[j]))
good.
good.
good.
good.
good.

File: 3563.lexicographically-smallest-string-after-adjacent-removals.cpp
//
// @lc app=leetcode id=3563 lang=cpp
//
// [3563] Lexicographically Smallest String After Adjacent Removals
//

// @lc code=start
class Solution {
public:
    bool isAdjacent(char a, char b) {
        int diff = abs(a - b);
        return diff == 1 || diff == 25;
    }
    
    string lexicographicallySmallestString(string s) {
        int n = s.size();
        if (n == 0) return "";
        
        // can[i][j] = true if s[i..j] can be completely removed
        vector<vector<bool>> can(n, vector<bool>(n, false));
        
        // Base case: length 2
        for (int i = 0; i + 1 < n; i++) {
            can[i][i+1] = isAdjacent(s[i], s[i+1]);
        }
        
        // Fill for increasing lengths (only even lengths >= 4 matter)
        for (int len = 4; len <= n; len += 2) {
            for (int i = 0; i + len - 1 < n; i++) {
                int j = i + len - 1;
                // Option 1: s[i] pairs with s[j], and middle is removable
                if (isAdjacent(s[i], s[j]) && can[i+1][j-1]) {
                    can[i][j] = true;
                    continue;
                }
                // Option 2: split into two removable parts
                for (int k = i + 1; k < j; k += 2) {
                    if (can[i][k] && can[k+1][j]) {
                        can[i][j] = true;
                        break;
                    }
                }
            }
        }
        
        // dp[i] = lexicographically smallest string from s[i..n-1]
        vector<string> dp(n + 1);
        dp[n] = "";
        
        for (int i = n - 1; i >= 0; i--) {
            bool emptyPossible = (n - i) % 2 == 0 && can[i][n-1];
            
            if (emptyPossible) {
                dp[i] = "";
            } else {
                // Initialize with s[i] + dp[i+1]
                dp[i] = string(1, s[i]) + dp[i+1];
                
                // Try removing prefixes of even length
                for (int j = i + 2; j < n; j += 2) {
                    // Try to remove s[i..j-1] and start with s[j]
                    if (can[i][j-1]) {
                        string candidate = string(1, s[j]) + dp[j+1];
                        if (candidate < dp[i]) {
                            dp[i] = candidate;
                        }
                    }
                }
            }
        }
        
        return dp[0];
    }
};
// @lc code=end